quarkchain/cluster/miner.py

```diff
@@ -227,51 +225,87 @@ def _mine_new_block_async(self):
                 return
 
             instance.process = AioProcess(
-                target=instance.mine_func,
-                args=(block, instance.input_q, instance.output_q, mining_params),
+                target=instance._mine_loop,
+                args=(
+                    instance.consensus_type,
+                    block,
+                    instance.input_q,
+                    instance.output_q,
+                    mining_params,
+                ),
             )
             instance.process.start()
             await handle_mined_block(instance)
 
-        if not self.enabled:
+        # no-op if enabled or mining remotely
+        if not self.enabled or self.remote:
             return None
         return asyncio.ensure_future(mine_new_block(self))
 
-    @staticmethod
-    def simulate_mine(
-        block,
-        input_q: MultiProcessingQueue,
-        output_q: MultiProcessingQueue,
-        mining_params: Dict,
-    ):
-        Miner._mine_loop(block, input_q, output_q, mining_params, Simulate)
-
-    @staticmethod
-    def mine_ethash(
-        block: Union[MinorBlock, RootBlock],
-        input_q: MultiProcessingQueue,
-        output_q: MultiProcessingQueue,
-        mining_params: Dict,
-    ):
-        Miner._mine_loop(block, input_q, output_q, mining_params, Ethash)
+    async def get_work(self, now=None) -> Optional[Tuple[bytes, int, int]]:
+        if not self.remote:
+            raise ValueError("Should only be used for remote miner")
+
+        if now is None:  # clock open for mock
+            now = time.time()
+        # 5 sec interval magic number
+        if not self.current_work or now - self.current_work.header.create_time > 5:
+            block = await self.create_block_async_func()
+            self.current_work = block
+
+        header = self.current_work.header
+        header_hash = header.get_hash_for_mining()
+        # store in memory for future retrieval during work submission
+        self.work_map[header_hash] = self.current_work
+
+        # clean up worker map
+        # TODO: for now, same param as go-ethereum
+        self.work_map = {
+            h: b
+            for h, b in self.work_map.items()
+            if now - b.header.create_time < 7 * 12
+        }
+
+        return header_hash, header.height, header.difficulty
+
+    async def submit_work(self, header_hash: bytes, nonce: int, mixhash: bytes) -> bool:
+        if not self.remote:
+            raise ValueError("Should only be used for remote miner")
+
+        if header_hash not in self.work_map:
+            return False
+        block = self.work_map[header_hash]
+        header = copy.copy(block.header)
+        header.nonce, header.mixhash = nonce, mixhash
+        try:
+            validate_seal(header, self.consensus_type)
```

Comment:
validate_seal should be put in add_block_async_func?

Reply:
yes it's already been called when validating root block and minor block (`RootState.validate_block_header` and `ShardState.__validate_block` respectively). an additional validation is added here to make sure it's not spam

Reply:
...and `add_block_async_func` calls those validation functions